iroh-bytes/src/downloader.rs

```diff
@@ -454,540 +503,495 @@ struct Service<G: Getter, D: Dialer> {
-                let delay_key = occupied_entry.remove().delay_key;
-                self.scheduled_request_queue.remove(&delay_key);
-                download_removed = true;
+                no_more_intents = true;
+                occupied_entry.remove();
             }
         }
 
-        if download_removed && !self.is_needed(hash) {
-            self.providers.remove(hash)
-        }
-    }
-
-    /// Handle a [`Message::PeersHave`].
-    fn handle_nodes_have(&mut self, hash: Hash, nodes: Vec<NodeInfo>) {
-        // check if this still needed
-        if self.is_needed(hash) {
-            self.providers.add_nodes(hash, &nodes);
+        if let Some(callbacks) = self.intents.remove(&intent_id) {
+            callbacks.on_finish.send(Err(DownloadError::Cancelled)).ok();
+            if let Some(sender) = callbacks.on_progress {
+                self.progress_tracker.unsubscribe(&kind, &sender);
+                sender
+                    .send(DownloadProgress::Abort(
+                        anyhow::Error::from(DownloadError::Cancelled).into(),
+                    ))
+                    .await
+                    .ok();
+            }
         }
-    }
-
-    /// Checks if this hash is needed.
-    fn is_needed(&self, hash: Hash) -> bool {
-        let as_blob = DownloadKind::Blob { hash };
-        let as_hash_seq = DownloadKind::HashSeq { hash };
-        self.current_requests.contains_key(&as_blob)
-            || self.scheduled_requests.contains_key(&as_blob)
-            || self.current_requests.contains_key(&as_hash_seq)
-            || self.scheduled_requests.contains_key(&as_hash_seq)
-    }
 
-    /// Check if this hash is currently being downloaded.
-    fn is_current_request(&self, hash: Hash) -> bool {
-        let as_blob = DownloadKind::Blob { hash };
-        let as_hash_seq = DownloadKind::HashSeq { hash };
-        self.current_requests.contains_key(&as_blob)
-            || self.current_requests.contains_key(&as_hash_seq)
-    }
-
-    /// Remove a hash from the scheduled queue.
-    fn unschedule(&mut self, hash: Hash) -> Option<(DownloadKind, PendingRequestInfo)> {
-        let as_blob = DownloadKind::Blob { hash };
-        let as_hash_seq = DownloadKind::HashSeq { hash };
-        let info = match self.scheduled_requests.remove(&as_blob) {
-            Some(req) => Some(req),
-            None => self.scheduled_requests.remove(&as_hash_seq),
-        };
-        if let Some(info) = info {
-            let kind = self.scheduled_request_queue.remove(&info.delay_key);
-            let kind = kind.into_inner();
-            Some((kind, info))
-        } else {
-            None
+        if no_more_intents {
+            if let Entry::Occupied(occupied_entry) = self.active_requests.entry(kind) {
+                occupied_entry.remove().cancellation.cancel();
+            } else {
+                self.queue.remove(&kind);
+            }
+            self.providers.remove_hash(&kind.hash());
```

Comment:
However unlikely, won't this remove the providers for the other kind if present?

Reply:
Good catch, added a check.